tests/src/deviceLib/hipAsynchronousStreams.cpp

```diff
@@ -0,0 +1,707 @@
+
+extern "C" __device__ __ockl_as_status_t
+__ockl_as_write_block(__ockl_as_stream_t ATTR_GLOBAL *stream, uchar service_id,
+                      ulong *connection_id, const uchar *str, uint32_t len,
+                      uchar flags);
+
+__device__ __ockl_as_status_t
+__hip_as_write_block(__ockl_as_stream_t *stream, uchar service_id,
+                     ulong *connection_id, const uchar *str, uint32_t len,
+                     uchar flags)
+{
+    __ockl_as_stream_t ATTR_GLOBAL *gstream =
+        reinterpret_cast<__ockl_as_stream_t ATTR_GLOBAL *>(stream);
+    return __ockl_as_write_block(gstream, service_id, connection_id, str, len,
+                                 flags);
+}
+
+/* END DUPLICATION */
+
+static __ockl_as_stream_t *
+createStream(void *ptr, uint buffer_size, uint num_packets)
+{
+    memset(ptr, 0, buffer_size);
+    __ockl_as_stream_t *r = (__ockl_as_stream_t *)ptr;
+    r->base_address = (__ockl_as_packet_t *)(&r[1]);
+    r->doorbell_signal = {0};
+    r->size = num_packets;
+
+    return r;
+}
+
+static uint8_t
+get_header_field(uint32_t header, uint8_t offset, uint8_t size)
+{
+    return (header >> offset) & ((1 << size) - 1);
+}
+
+static uint8_t
+get_packet_type(uint32_t header)
+{
+    return get_header_field(header, __OCKL_AS_PACKET_HEADER_TYPE,
+                            __OCKL_AS_PACKET_HEADER_WIDTH_TYPE);
+}
+
+static uint8_t
+get_packet_flags(uint32_t header)
+{
+    return get_header_field(header, __OCKL_AS_PACKET_HEADER_FLAGS,
+                            __OCKL_AS_PACKET_HEADER_WIDTH_FLAGS);
+}
+
+static uint8_t
+get_packet_bytes(uint32_t header)
+{
+    return get_header_field(header, __OCKL_AS_PACKET_HEADER_BYTES,
+                            __OCKL_AS_PACKET_HEADER_WIDTH_BYTES);
+}
+
+static uint8_t
+get_packet_service(uint32_t header)
+{
+    return get_header_field(header, __OCKL_AS_PACKET_HEADER_SERVICE,
+                            __OCKL_AS_PACKET_HEADER_WIDTH_SERVICE);
+}
+
+const unsigned int __OCKL_AS_PACKET_SIZE = sizeof(__ockl_as_packet_t);
+
+/* END DUPLICATION */
+
+using namespace std;
+
+#define STR_HELLO_WORLD "hello world"
+#define STRLEN_HELLO_WORLD 11
+
+const unsigned int THREADS_PER_BLOCK = 123; // include a partial warp
+const unsigned int NUM_BLOCKS = 3; // because powers of two are too convenient
+const unsigned int NUM_THREADS = NUM_BLOCKS * THREADS_PER_BLOCK;
+const unsigned int NUM_PACKETS_INSUFFICIENT = NUM_THREADS - 23;
+const unsigned int NUM_PACKETS_LARGE = NUM_THREADS * 4;
+const unsigned int NUM_SERVICES = 7;
+
+unsigned int
+read_uint(const unsigned char *ptr)
+{
+    unsigned int value = 0;
+
+    for (int ii = sizeof(unsigned int) - 1; ii >= 0; --ii) {
+        value <<= 8;
+        value |= ptr[ii];
+    }
+
+    return value;
+}
+
+__global__ void
+singlePacketSingleProducer(__ockl_as_stream_t *stream)
+{
+    uint len = STRLEN_HELLO_WORLD;
+
+    uint64_t connection_id;
+
+    __hip_as_write_block(stream, 42, &connection_id,
```

Comment:
42 maybe the answer to the Life, Universe and Everything, but not sure what it is doing here. Maybe you wanted ```__OCKL_AS_BUILTIN_SERVICE_PRINTF```?

Reply:
The PRINTF enum is not required here. I have removed it entirely now. Also replaced the magic number 42 with a const integer TEST_SERVICE. It just needs to be some non-zero number for the test to check successful transmission.